crates/wasmparser/src/validator/types.rs

```diff
@@ -140,58 +139,651 @@ fn push_primitive_wasm_types(ty: &PrimitiveValType, lowered_types: &mut LoweredT
     }
 }
 
-/// Represents a unique identifier for a type known to a [`crate::Validator`].
+/// A trait shared by all type identifiers.
+///
+/// Any id that can be used to get a type from a `Types`.
+//
+// Or, internally, from a `TypeList`.
+pub trait TypeIdentifier:
+    TypeIdentifierImpl<DataImpl = Self::Data> + std::fmt::Debug + Copy + Eq + Sized + 'static
+{
+    /// The data pointed to by this type of id.
+    type Data: TypeData<Id = Self>;
+}
+
+/// A trait shared by all types within a `Types`.
+///
+/// This is the data that can be retreived by indexing with the associated
+/// [`TypeIdentifier`].
+pub trait TypeData: TypeDataImpl + std::fmt::Debug {
+    /// The identifier for this type data.
+    type Id: TypeIdentifier<Data = Self>;
+}
+
+/// A type that can be aliased in the component model.
+pub trait Aliasable: AliasableImpl {}
```

Comment:
Wasmparser has a breaking major version bump all the time, so I think it's fine to skip the `sealed` bits perhaps and go straight to just making everything `pub`. If we change it then it's not much different than any other breaking change in `wasmparser` sort of

Reply:
I guess I can make it `pub` but `doc(hidden)` at least so that it doesn't clutter the API docs and confuse people.